File: Study Plan/LeetCode 75/933 Number of Recent Calls/NumberOfRecentCalls.cs
/*  
    Problem
    --------------------------------------------------------------------------------
    933. Number of Recent Calls
*/

/*  
    Solution
    --------------------------------------------------------------------------------
    - Runtime 225 ms (Beats 100.00% of users with C#)
    - Memory 102.04 MB (Beats 11.80% of users with C#)
*/

public class RecentCounter
{
    // Represents a first-in, first-out (FIFO) collection of objects.
    private readonly Queue<int> requests;

    public RecentCounter()
    {
        requests = new Queue<int>();
    }

    public int Ping(int t)
    {
        // .Enqueue(T) adds an object to the end of the Queue<T>.
        requests.Enqueue(t);

        // .Peek() returns the object at the beginning of the Queue<T>
        // without removing it.
        while (requests.Peek() < t - 3000)
        {
            // .Dequeue() removes and returns the object at the beginning
            // of the Queue<T>.
            requests.Dequeue();
        }
        
        return requests.Count;
    }
}

/**
 * Your RecentCounter object will be instantiated and called as such:
 * RecentCounter obj = new RecentCounter();
 * int param_1 = obj.Ping(t);
 */

File: Study Plan/LeetCode 75/933 Number of Recent Calls/NumberOfRecentCalls.java
/*  
    Problem
    --------------------------------------------------------------------------------
    933. Number of Recent Calls
*/

/*  
    Solution
    --------------------------------------------------------------------------------
    - Runtime 22 ms (Beats 48.70% of users with Java)
    - Memory 55.82 MB (Beats 6.59% of users with Java)

*/

public class RecentCounter {
    // Represents a first-in, first-out (FIFO) collection of objects.
    private final Deque<Integer> requests;

    public RecentCounter() {
        requests = new ArrayDeque<>();
    }

    public int ping(int t) {
        // .offer() adds an object to the end of the Queue.
        requests.offer(t);
        // .peek() returns the object at the beginning of the Queue
        // without removing it.
        while (!requests.isEmpty() && requests.peek() < t - 3000) {
            // .poll() removes and returns the object at the beginning
            // of the Queue.
            requests.poll();
        }
        return requests.size();
    }
}

/**
 * Your RecentCounter object will be instantiated and called as such:
 * RecentCounter obj = new RecentCounter();
 * int param_1 = obj.ping(t);
 */